### PR TITLE
SYCL outer TeamPolicy parallel_reduce

### PR DESCRIPTION
**Pull request by masterleinad**
Depends on #3783. With this and #3815 TeamPolicy should be feature-complete for `SYCL` (excluding `vector_size>1`).

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/SYCL/Kokkos_SYCL_Team.hpp [hunk 1]
```diff
@@ -109,10 +109,17 @@ class SYCLTeamMember {
...
 
   template <class ValueType>
-  KOKKOS_INLINE_FUNCTION void team_broadcast(ValueType& /*val*/,
-                                             const int& /*thread_id*/) const {
-    // FIXME_SYCL
-    Kokkos::abort("Not implemented!");
+  KOKKOS_INLINE_FUNCTION void team_broadcast(ValueType& val,
+                                             const int& thread_id) const {
+    // Wait for shared data write until all threads arrive here
+    m_item.barrier(sycl::access::fence_space::local_space);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Team.hpp:115]: ...and my answer in https://github.com/kokkos/kokkos/pull/3815#discussion_r590628347. :slightly_smiling_face:

File: core/unit_test/TestTeamReductionScan.hpp [hunk 2]
```diff
@@ -48,6 +48,8 @@
 
 namespace Test {
 
+// FIXME_SYCL needs team_scan
+#ifndef KOKKOS_ENABLE_SYCL
```
Comment (dalg24) [inline core/unit_test/TestTeamReductionScan.hpp:52]: Does this work with #3815?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 3]
```diff
@@ -556,12 +529,281 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
+    // necessary.
+    if (m_result_ptr) {
+      Kokkos::Impl::DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+                             Kokkos::Experimental::SYCLDeviceUSMSpace>(
+          space, m_result_ptr, results_ptr,
+          sizeof(*m_result_ptr) * value_count);
+      space.fence();
+    }
+
+    sycl::free(results_ptr, q);
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:741]: You forgot to free `results_ptr2`.  Put these in a `std::unique_ptr` with a custom deleter.

**Review by dalg24 — DISMISSED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:741]: Bummer. I thought I caught all of them in the various versions. :slightly_frowning_face:

File: core/unit_test/TestTeamReductionScan.hpp (hunk 2 above)
Comment (masterleinad) [reply core/unit_test/TestTeamReductionScan.hpp:52]: Well, I need both pull requests for that. :slightly_smiling_face:

Comment (masterleinad) [reply core/unit_test/TestTeamReductionScan.hpp:52]: I'll look into it but consider it part of #3815 so I would appreciate if we can move forward here independently.

Comment (masterleinad) [reply core/unit_test/TestTeamReductionScan.hpp:52]: This works now in #3815 after rebasing there.

File: core/unit_test/CMakeLists.txt [hunk 4]
```diff
@@ -44,7 +44,7 @@ SET(KOKKOS_OPENMPTARGET_FEATURE_LEVEL 13)
 SET(KOKKOS_OPENMPTARGET_NAME Experimental::OpenMPTarget)
 SET(KOKKOS_SERIAL_FEATURE_LEVEL 999)
 SET(KOKKOS_SERIAL_NAME Serial)
-SET(KOKKOS_SYCL_FEATURE_LEVEL 13)
+SET(KOKKOS_SYCL_FEATURE_LEVEL 999)
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:47]: 🥳

**Review by crtrott — COMMENTED** [review]
Almost there. Just one question on the queue capture.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 5]
```diff
@@ -556,12 +529,289 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
+
+    // FIXME_SYCL optimize
+    const size_t wgroup_size = m_team_size;
+    std::size_t size         = m_league_size * m_team_size;
+    const auto init_size =
+        std::max<std::size_t>((size + wgroup_size - 1) / wgroup_size, 1);
+    const unsigned int value_count =
+        FunctorValueTraits<ReducerTypeFwd, WorkTagFwd>::value_count(
+            selected_reducer);
+    auto deleter = [q](pointer_type ptr) { sycl::free(ptr, q); };
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:564]: do you mean to capture by copy?  above you declare it as a non-const reference

**Review by crtrott — APPROVED** [review]

**Pull request merged**